core/state/state_object.go

```diff
@@ -245,10 +251,34 @@ func (self *StateObject) Value() *big.Int {
 	panic("Value on StateObject should never be called")
 }
 
+func (self *StateObject) ForEachStorage(full bool, cb func(key, value common.Hash) bool) {
+	// When iterating over the storage check the cache first
+	for h, sv := range self.storage {
+		if sv.dirty || full {
+			cb(h, sv.value)
+		}
+	}
+
+	// if full storage retrieval is disabled abort here
+	if !full {
+		return
+	}
+
+	it := self.trie.Iterator()
+	for it.Next() {
+		// ignore cached values
+		key := common.BytesToHash(self.trie.GetKey(it.Key))
+		if _, ok := self.storage[key]; !ok {
+			cb(key, common.BytesToHash(it.Value))
+		}
+	}
+}
```

Comment:
Hmm, this method will return duplicates if full is requested for both modified as well as requested keys. Shouldn't there rather be two separate sections?

```go
if !full {
	for h, sv := range self.storage {
 		if sv.dirty {
 			cb(h, sv.value)
 		}
	}
	return
}
it := self.trie.Iterator()
	for it.Next() {
		// ignore cached values
		key := common.BytesToHash(self.trie.GetKey(it.Key))
		if _, ok := self.storage[key]; !ok {
			cb(key, common.BytesToHash(it.Value))
		}
	}
```